src/render/View3D.cpp

```diff
@@ -498,6 +514,42 @@ void View3D::paintGL()
     glCheckError();
 }
 
+
+QImage View3D::renderScreenShot(int width, int height)
+{
+    if (m_badOpenGL)
+        return QImage();
+
+    double dPR = getDevicePixelRatio();
+    if (width < 0)
+        width = this->width();
+
+    if (height < 0)
+        height = this->height();
+
+    // Note: the reason for the following two lines and the subsequent downsampling of the
+    //       resulting image is to make the radius of rendered points consistent between the
+    //       gui and screenshot on displas with dPR != 1.
+    width *= dPR;
+    height *= dPR;
+
+    QOpenGLFramebufferObject framebuffer(QSize(width, height), QOpenGLFramebufferObject::Depth);
+    framebuffer.bind();
+    resizeViewport(width, height);
```

Comment:
Ah, of course!

The problem here is that in the vertex shader `gl_PointSize` must be specified in pixels, so the shader needs to know the size of the viewport.  This is communicated with the `pointPixelScale` uniform variable (see `View3d::drawPoints()`).  Just looking at that code, it currently assumes `dPR` should be multiplied in, and grabs information from the camera directly.  This is another thing which should be disentangled to make it possible to cleanly render to an offscreen buffer: everything about the camera and view should be passed along explicitly in `transState`.  Luckily transState already has `viewSize`, so you can probably just use something like

```C++
prog.setUniformValue("pointPixelScale", (GLfloat)(0.5*transState.viewSize.x*transState.projMatrix[0][0]));
```